Draft of purrr walk post (#640)
tidyverse/tidyverse.org: content/blog/purrr-walk-this-way/index.Rmd

|                       | List     | Atomic            | Same type   | Nothing   |
|-----------------------|----------|-------------------|-------------|-----------|
| One argument          | `map()`  | `map_lgl()`, ...  | `modify()`  | `walk()`  |
| Two arguments         | `map2()` | `map2_lgl()`, ... | `modify2()` | `walk2()` |
| One argument + index  | `imap()` | `imap_lgl()`, ... | `imodify()` | `iwalk()` |
| N arguments           | `pmap()` | `pmap_lgl()`, ... | ---         | `pwalk()` |

## What's up with `walk()`?

Based on the table above, you might think that `walk()` isn't very useful.
Indeed, `walk()`, `walk2()`, and `pwalk()` all invisibly return `.x`.
However, they come in handy when you want to call a function for its ***side effects*** rather than its return value.

Here, we'll go through two common use cases: saving multiple CSVs, and multiple plots.
We'll also make use of the [fs](https://fs.r-lib.org/) package, a cross-platform interface to file system operations, to inspect our outputs.

> If you want to try this out but don't want to save files locally, there's a [companion project on **Posit Cloud**](https://posit.cloud/content/5983147) where you can follow along.

Reviewer: Would it be better to use `.qmd` files on Cloud? Can we make the project open with one of them already open?

Author: Done.

```{r}
#| label: try-it
#| echo: false
blogdown::shortcode("test-drive-cloud", url = "https://posit.cloud/content/5983147")
```

## Writing (and deleting) multiple CSVs

To get started, we'll need some data.
Let's use the [gapminder](https://googlesheets4.tidyverse.org/reference/gs4_examples.html) example Sheet built into [googlesheets4](https://googlesheets4.tidyverse.org/).
Because there are multiple worksheets (one for each continent), we'll use `map()` to apply `read_sheet()`[^1] to each one, and get back a list of data frames.

```{r}
#| label: libs
#| message: false
library(tidyverse)
library(googlesheets4)
```

```{r}
#| label: read-sheets
# since we're using data built into googlesheets4 we don't need to auth

> gs4_deauth()

Reviewer: Can you hide these lines from the blog post?

ss <- gs4_example("gapminder") # get sheet id
sheets <- sheet_names(ss) # get the names of individual sheets
gap_dfs <- map(sheets, .f = \(x) read_sheet(ss, sheet = x))
```

> Note that the backslash syntax for anonymous functions (e.g. `\(x) x +1`) was introduced in base R version 4.1.0. If you're using an earlier version of R, you can use a formula instead (e.g. `~ .x + 1`).

Reviewer: ```suggestion
Note that the backslash syntax for anonymous functions (e.g. `\(x) x + 1`) was introduced in base R version 4.1.0 as a shorthand for `function(x) x + 1`. If you're using an earlier version of R, you can use purrr's shorthand: a formula  (e.g. `~ .x + 1`).
```

> Typically, you'd want to combine these data frames into one for the purposes of working with the data in R.

Reviewer: ```suggestion
Typically, you'd want to combine these data frames into one to make it easier to work with your data.
```

To do so, we'll use `list_rbind()` on `gap_dfs`.
I've kept the intermediary object, since we'll use it in a moment with `walk()`, but could have just as easily piped the output directly.

> The combination of [`purrr::map()` and `list_rbind()`](https://r4ds.hadley.nz/iteration.html?#purrrmap-and-list_rbind) is a handy one you can learn more about in the linked section of R for Data Science.
> 

Reviewer: ```suggestion
The combination of `purrr::map()` and `list_rbind()` is a handy one that you can learn more about in the [R for Data Science](https://r4ds.hadley.nz/iteration.html?#purrrmap-and-list_rbind).

```

Author: The awkward wording was my attempt to have the link text be descriptive, but I agree this is better.

```{r}
#| label: combine-dfs
gap_combined <- gap_dfs |>
  list_rbind()
```

Now let's say that, for whatever reason, you'd like to save the data from these sheets as individual CSVs.
This is where `walk()` comes into play---writing out the file with `write_csv()` is a "side effect." We'll use [`fs::dir_create()`](https://fs.r-lib.org/reference/create.html) to create a data folder to put our files into[^2], and build a vector of paths/file names.
Since we have two arguments, the list of data frames, and the paths, we'll use `walk2()`.

```{r}
#| label: write-csvs
fs::dir_create("data")
paths <- str_glue("data/gapminder_{tolower(sheets)}.csv")

> walk2(gap_dfs, paths, write_csv)

Reviewer: Maybe use `\(df, name) write_csv(df, name)` to make the arguments a bit clearer?

```

> To see what we've done, we can use [`fs::dir_tree()`](https://fs.r-lib.org/reference/dir_tree.html) to see the contents of the directory as a tree, or [`fs::dir_ls()`](https://fs.r-lib.org/reference/dir_ls.html) to return the paths as a list.

Reviewer: ```suggestion
To see what we've done, we can use [`fs::dir_tree()`](https://fs.r-lib.org/reference/dir_tree.html) to see the contents of the directory as a tree, or [`fs::dir_ls()`](https://fs.r-lib.org/reference/dir_ls.html) to return the paths as a vector.
```

```{r}
#| label: fs-view
fs::dir_tree("data")
fs::dir_ls("data")

> ```

Reviewer: I do think it's worth mentioning `glob`/`regexp` here

```{r}
#| label: conditional-bars
# modified conditional bars function from R4DS
conditional_bars <- function(df, condition, var) {
  df |>
    filter({{ condition }}) |>
    ggplot(aes(x = {{ var }})) +
    geom_bar() +
    ggtitle(rlang::englue("Count of diamonds by {{var}} where {{condition}}"))
}
```

It's easy enough to run this for one condition, for example for the diamonds with `cut == "Good"`.

```{r}
#| label: goodclarity
#| fig-alt: "Bar chart showing count of diamonds by clarity in the diamonds dataset where the cut == Good."
diamonds |> conditional_bars(cut == "Good", clarity)
```

But what if we want to make a and save plot for each cut?
Again, it's `map()` and `walk()` to the rescue.

Because we're using the same data (`diamonds`) and conditioning on the same variable (`cut`), we'll only need to `map()` across the levels of `cut`, and can hard code the rest into the anonymous function.

```{r}
#| label: map-plots
# get the levels
cuts <- levels(diamonds$cut)

# make the plots
plots <- map(
  cuts, 
  \(x) conditional_bars(df = diamonds, cut == {{ x }}, clarity)

>   )

Reviewer: All on one line?

```

> 

Reviewer: I think it'd be worth a note commenting that it's just fine to store ggplots in a list.

As we did when saving our CSVs, we'll use fs to create a directory to store them in, and make a vector of paths for file names.

```{r}
#| label: plot-paths
# make the folder to put them it (if exists, {fs} does nothing)
fs::dir_create("plots")
# make the file names
plot_paths <- str_glue("plots/{tolower(cuts)}_clarity.png")
```

Now we can use the paths and plots with `walk2()` to pass them as arguments to [`ggsave()`](https://ggplot2.tidyverse.org/reference/ggsave.html).

> Note that, rather than putting constant arguments (such as height and width) in `…`, we pass them directly into `ggsave()` in an anonymous function.

Reviewer: I think you should drop this sentence by using anonymous functions everywhere.
